### ISSUE TITLE
Update removes defined exac version

### ISSUE THREAD

**Issue by muuvmuuv** [pnpm/pnpm#8240]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

9.x

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI

### Reproduction steps

```shell
set TS_VERSION $(cat node_modules/@angular-devkit/build-angular/package.json | jq '.peerDependencies.typescript' | xargs)
pnpm remove typescript # required for save-exac
pnpm add --save-dev --save-exac typescript@"$TS_VERSION"
pnpm update
```

### Describe the Bug

When trying the save a version of a package with an exac version range (like above), `pnpm update` will resolve that range in "^..." which breaks peer dependencies.

Besides that, as you see above, currently I need to remove and add a package if I want to update/install with a exact specified range. I do that, because I have written an automated updater script for my Angular projects (sometimes TS isn't set to the correct version).

### Expected Behavior

Update should not modify version range modifier except with flag `--latest`.

### Which Node.js version are you using?

20.x

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by zkochan** [comment]
I could not reproduce the bug. After running `pnpm add --save-dev --save-exac typescript@"$TS_VERSION"` was typescript added with an exact version to `package.json`?

**Comment by muuvmuuv** [comment]
Yes, IMO it should be added as the way as I defined it and after doing update should leave it as is. This destroys the hole point of semver.

![SCR-20240712-jbhf](https://github.com/user-attachments/assets/2077271b-7ef6-406b-b746-e54269965a08)
![SCR-20240712-jbih](https://github.com/user-attachments/assets/25ffb066-bdd1-4be7-b08d-434aa44d7020)

**Comment by zkochan** [comment]
The problem is not with the update command. The problem is with the add command, which added `typescript` with version range `>=5.4 <5.6` into `package.json` instead of an exact version.

**Comment by muuvmuuv** [comment]
But I want it to be ">=5.4 <5.6" in my package.json. The way pnpm updated it will potentially install a completely wrong version on the next update. It must stay at that range anytime. There is no reason to edit my version range definition if not directly said to do that. Especially when there is the flag `--save-exac` but not respected by `update`...

**Comment by muuvmuuv** [comment]
What is the alternative? It is not about add but update, update does change the semver definition afterwards and breaks compatibility with other packages.

**Comment by muuvmuuv** [comment]
I want to pick that up again since I think the main point is still not clear and "--save-exact" was just a test.

1. `pnpm init`
2. `pnpm add "typescript@>=5.5 <5.6"`
3. Correctly added range to package.json
4. `pnpm update`
5. Changes ">=5.5 <5.6" to "^5.6.2" in package.json

The issue with the "^5.6.2" is that now when I do an update again in a few month, it e.g. install "5.13.2" which will potentially introduce breaking api changes to other libs that depend on typescript. Instead if it would leave my defined range as-is it will always update the lockfile to the next publish release within my defined range. Which could e.g. update from "5.6.1" → "5.6.2".

I just say that this is not ideal to override the users choice given in package.json file and it is not needed either. The semver already says what to install and the actual version is locked in the lockfile.

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

Verified on pnpm 12.5.1: `pnpm update` and `pnpm update --latest` keep an exact version exact.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
